pallets/registrar/src/lib.rs

```diff
@@ -358,9 +358,12 @@ pub mod pallet {
             // Account for on_finalize weight
             let mut weight = Weight::zero().saturating_add(T::DbWeight::get().reads(1));
 
-            let buffered_paras = BufferedParasToDeregister::<T>::get();
-            for _ in 0..buffered_paras.len() {
+            let buffered_paras = BufferedParasToDeregister::<T>::take();
+
+            for para_id in buffered_paras {
                 weight += T::InnerRegistrar::deregister_weight();
+                // Deregister (in the relay context) each paraId present inside the buffer
+                T::InnerRegistrar::deregister(para_id);
```

Comment:
we cannt. Because `initializer_on_new_session` happens `on_initialize` of the new session block, and the relay registrars pallet finishes downgrading a parachain to a parathread `on_finalize` of the same block.

That is why we need to buffer it to be executed `on_initialize` of the block posterior to the session change block

Reply:
But this buffer is executed on_finalize in the same block, not on_initialize of the next block. I guess it's the same because of pallet order

Reply:
Oh sorry, you changed it. Then also change the comment above, remove `// Account for on_finalize weight` and add a write for take()

Reply:
fixed